app/controllers/concerns/accounts/authorization.rb

```diff
@@ -0,0 +1,185 @@
+#
+# See COPYRIGHT and LICENSE files for more details.
+# ++
+
+##
+# Intended to be used by the ApplicationController to provide authorization helpers
+module Accounts::Authorization
+  extend ActiveSupport::Concern
+
+  METHODS_ENFORCING_AUTHORIZATION = %i[require_admin authorize authorize_global authorize_in_optional_project].freeze
+
+  included do
+    class_attribute :authorization_ensured,
+                    default: {
+                      only: [],
+                      except: [],
+                      generally_allowed: false,
+                      controller: self
+                    }
+  end
+
+  private
+
+  def authorization_check_required
+    unless authorization_is_ensured?(params[:action])
+      if Rails.env.development?
+        raise <<-MESSAGE
+          Authorization check required for #{params[:action]} in #{self.class.name}.
+
+          Use any method of 'authorize', 'authorize_global', 'authorize_in_optional_project'
+          or 'require_admin' to ensure authorization. If authorization is checked by any other means,
+          affirm the same by calling 'authorization_checked!' in the controller. If the authorization does
+          not need to be checked for this action, affirm the same by calling 'no_authorization_required!'
+        MESSAGE
+      else
+        render_403
+      end
+    end
+  end
+
+  # Authorize the user for the requested controller action.
+  # To be used in before_action hooks
+  def authorize(ctrl = params[:controller], action = params[:action])
+    do_authorize({ controller: ctrl, action: }, global: false)
+  end
+
+  # Authorize the user for the requested controller action outside a project
+  # To be used in before_action hooks
+  def authorize_global
+    action = { controller: params[:controller], action: params[:action] }
+    do_authorize(action, global: true)
+  end
+
+  # Find a project based on params[:project_id]
+  def authorize_in_optional_project
+    @project = Project.find(params[:project_id]) if params[:project_id].present?
+
+    do_authorize({ controller: params[:controller], action: params[:action] }, global: params[:project_id].blank?)
+  rescue ActiveRecord::RecordNotFound
+    render_404
+  end
+
+  # Deny access if user is not allowed to do the specified action.
+  #
+  # Action can be:
+  # * a parameter-like Hash (eg. { controller: '/projects', action: 'edit' })
+  # * a permission Symbol (eg. :edit_project)
+  def do_authorize(action, global: false) # rubocop:disable Metrics/PerceivedComplexity
+    is_authorized = if global
+                      User.current.allowed_based_on_permission_context?(action)
+                    else
+                      User.current.allowed_based_on_permission_context?(action,
+                                                                        project: @project || @projects,
+                                                                        entity: @work_package || @work_packages)
```

Comment:
I'm just not feeling adventurous enough. There is a slight difference in how explicit the statements are. E.g. with `authorize_global` it would be ok to have an instance variable called `@work_package` without it getting factored into the authorization check. 

I really dislike the way before_actions rely on instance variable to pass around data, but that is what we have.